File: src-winui/src/GSBT.Core/Services/BackupRetentionVerifier.cs
using GSBT.Core.Common;

namespace GSBT.Core.Services;

/// <summary>
/// Checks whether GSBT retention backup folders still exist under the configured default backup root
/// (same layout as <see cref="SaveFolderBackupService"/>).
/// </summary>
public static class BackupRetentionVerifier
{
    /// <returns>True if at least one non-empty retention run exists: a directory matching <c>{sanitizedName} - Backup *</c> containing at least one file (recursive).</returns>
    public static bool HasRetentionArtifact(string backupRoot, string gameName, bool subfolderPerGame)
    {
        if (string.IsNullOrWhiteSpace(backupRoot) || !Directory.Exists(backupRoot))
        {
            return false;
        }

        var safe = GameNameInputValidation.SanitizeForWindowsPathSegment(
            string.IsNullOrWhiteSpace(gameName) ? "Game" : gameName);

        var baseDir = subfolderPerGame
            ? Path.Combine(backupRoot, safe)
            : backupRoot;

        if (!Directory.Exists(baseDir))
        {
            return false;
        }

        if (HasRetentionRunDirectory(baseDir, safe) || HasLegacyFlatRegExport(baseDir, safe))
        {
            return true;
        }

        return false;
    }

    /// <summary>Infers the newest retention backup time (prefers AppData checkpoint <c>checkpointCapturedAtUtc</c> when present).</summary>
    public static string? TryInferLatestLastBackupIso(
        string backupRoot,
        string gameName,
        bool subfolderPerGame)
    {
        var latest = TryGetLatestRetentionRunDirectory(backupRoot, gameName, subfolderPerGame);
        DateTime bestUtc = DateTime.MinValue;
        if (!string.IsNullOrWhiteSpace(latest))
        {
            bestUtc = EffectiveUtcForRetentionRun(latest);
        }

        var safe = GameNameInputValidation.SanitizeForWindowsPathSegment(
            string.IsNullOrWhiteSpace(gameName) ? "Game" : gameName);
        var baseDir = subfolderPerGame
            ? Path.Combine(backupRoot, safe)
            : backupRoot;

        if (Directory.Exists(baseDir))
        {
            var legacyUtc = TryGetLatestLegacyFlatRegUtc(baseDir, safe);
            if (legacyUtc is { } lu && lu > bestUtc)
            {
                bestUtc = lu;
            }
        }

        if (bestUtc == DateTime.MinValue)
        {
            return null;
        }

        return bestUtc.ToString("O");
    }

    /// <summary>Full path of the newest non-empty retention run for <paramref name="gameName"/> (by effective checkpoint/write time).</summary>
    public static string? TryGetLatestRetentionRunDirectory(string backupRoot, string gameName, bool subfolderPerGame)
    {
        if (string.IsNullOrWhiteSpace(backupRoot) || !Directory.Exists(backupRoot))
        {
            return null;
        }

        var safe = GameNameInputValidation.SanitizeForWindowsPathSegment(
            string.IsNullOrWhiteSpace(gameName) ? "Game" : gameName);

        var baseDir = subfolderPerGame
            ? Path.Combine(backupRoot, safe)
            : backupRoot;

        if (!Directory.Exists(baseDir))
        {
            return null;
        }

        string? best = null;
        var bestEff = DateTime.MinValue;
        foreach (var dir in EnumerateNonEmptyRetentionRunDirectories(baseDir, safe))
        {
            var eff = EffectiveUtcForRetentionRun(dir);
            if (eff > bestEff)
            {
                bestEff = eff;
                best = dir;
            }
        }

        return best;
    }

    private static DateTime EffectiveUtcForRetentionRun(string runDir)
    {
        if (BackupRunManifestStore.TryReadCheckpointCapturedAtUtc(runDir, out var cp))
        {
            return cp;
        }

        try
        {
            return Directory.GetLastWriteTimeUtc(runDir);
        }
        catch
        {
            return DateTime.MinValue;
        }
    }

    /// <summary>Sum of on-disk file lengths for every non-empty retention run folder for <paramref name="gameName"/>.</summary>
    public static long ComputeTotalRetentionBackupBytes(string backupRoot, string gameName, bool subfolderPerGame)
    {
        if (string.IsNullOrWhiteSpace(backupRoot) || !Directory.Exists(backupRoot))
        {
            return 0;
        }

        var safe = GameNameInputValidation.SanitizeForWindowsPathSegment(
            string.IsNullOrWhiteSpace(gameName) ? "Game" : gameName);

        var baseDir = subfolderPerGame
            ? Path.Combine(backupRoot, safe)
            : backupRoot;

        if (!Directory.Exists(baseDir))
        {
            return 0;
        }

        long sum = 0;
        foreach (var dir in EnumerateNonEmptyRetentionRunDirectories(baseDir, safe))
        {
            sum += BackupFolderSizeEstimator.ComputeDirectoryLogicalSize(dir);
        }

        foreach (var file in EnumerateLegacyFlatRegExports(baseDir, safe))
        {
            try
            {
                sum += new FileInfo(file).Length;
            }
            catch
            {
                // ignore unreadable files
            }
        }

        return sum;
    }

    private static bool HasRetentionRunDirectory(string baseDir, string safe)
    {
        foreach (var dir in EnumerateNonEmptyRetentionRunDirectories(baseDir, safe))
        {
            return true;
        }

        return false;
    }

    private static bool HasLegacyFlatRegExport(string baseDir, string safe)
    {
        foreach (var _ in EnumerateLegacyFlatRegExports(baseDir, safe))
        {
            return true;
        }

        return false;
    }

    private static IEnumerable<string> EnumerateNonEmptyRetentionRunDirectories(string baseDir, string safe)
    {
        var prefix = $"{safe} - Backup";
        foreach (var dir in Directory.EnumerateDirectories(baseDir))
        {
            if (!Path.GetFileName(dir).StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
            {
                continue;
            }

            if (RetentionRunContainsAtLeastOneFile(dir))
            {
                yield return dir;
            }
        }
    }

    private static IEnumerable<string> EnumerateLegacyFlatRegExports(string baseDir, string safe)
    {
        if (!Directory.Exists(baseDir))
        {
            yield break;
        }

        var prefix = $"{safe} - Backup";
        foreach (var file in Directory.EnumerateFiles(baseDir, $"{prefix}*.reg"))
        {
            if (Path.GetFileName(file).StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
            {
                yield return file;
            }
        }
    }

    private static DateTime? TryGetLatestLegacyFlatRegUtc(string baseDir, string safe)
    {
        DateTime? best = null;
        foreach (var file in EnumerateLegacyFlatRegExports(baseDir, safe))
        {
            try
            {
                var t = File.GetLastWriteTimeUtc(file);
                if (best is null || t > best.Value)
                {
                    best = t;
                }
            }
            catch
            {
                // ignore
            }
        }

        return best;
    }

    /// <summary>Empty retention folders (name matches but copy failed or user recreated shell dirs) do not count as backups.</summary>
    public static bool RetentionRunContainsAtLeastOneFile(string retentionRunDirectory)
    {
        if (string.IsNullOrWhiteSpace(retentionRunDirectory) || !Directory.Exists(retentionRunDirectory))
        {
            return false;
        }

        try
        {
            using var e = Directory.EnumerateFiles(retentionRunDirectory, "*", SearchOption.AllDirectories).GetEnumerator();
            return e.MoveNext();
        }
        catch
        {
            return false;
        }
    }
}

File: src-pyqt/core/backup_retention_verifier.py
"""Check whether GSBT retention backup folders exist under the configured backup root."""

from __future__ import annotations

import os
import re
from typing import Iterable, List, Optional

from core.game_name_validation import sanitize_for_windows_path_segment


def retention_run_contains_at_least_one_file(retention_run_directory: str) -> bool:
    if not retention_run_directory or not os.path.isdir(retention_run_directory):
        return False
    try:
        for _root, _dirs, files in os.walk(retention_run_directory):
            if files:
                return True
    except OSError:
        return False
    return False


def _enumerate_nonempty_retention_run_dirs(base_dir: str, safe: str) -> Iterable[str]:
    prefix = f"{safe} - Backup"
    if not os.path.isdir(base_dir):
        return
    try:
        for name in os.listdir(base_dir):
            full = os.path.join(base_dir, name)
            if not os.path.isdir(full):
                continue
            if not name.startswith(prefix):
                continue
            if retention_run_contains_at_least_one_file(full):
                yield full
    except OSError:
        return


def _enumerate_legacy_flat_reg_exports(base_dir: str, safe: str) -> Iterable[str]:
    prefix = f"{safe} - Backup"
    if not os.path.isdir(base_dir):
        return
    try:
        for name in os.listdir(base_dir):
            if not name.lower().endswith(".reg"):
                continue
            if name.startswith(prefix):
                yield os.path.join(base_dir, name)
    except OSError:
        return


def has_retention_artifact(backup_root: str, game_name: str, subfolder_per_game: bool) -> bool:
    if not backup_root or not os.path.isdir(backup_root):
        return False
    safe = sanitize_for_windows_path_segment(game_name)
    base_dir = os.path.join(backup_root, safe) if subfolder_per_game else backup_root
    if not os.path.isdir(base_dir):
        return False
    for _ in _enumerate_nonempty_retention_run_dirs(base_dir, safe):
        return True
    for _ in _enumerate_legacy_flat_reg_exports(base_dir, safe):
        return True
    return False


def try_get_latest_retention_run_directory(
    backup_root: str, game_name: str, subfolder_per_game: bool
) -> Optional[str]:
    if not backup_root or not os.path.isdir(backup_root):
        return None
    safe = sanitize_for_windows_path_segment(game_name)
    base_dir = os.path.join(backup_root, safe) if subfolder_per_game else backup_root
    if not os.path.isdir(base_dir):
        return None
    best: Optional[str] = None
    best_mtime = 0.0
    for d in _enumerate_nonempty_retention_run_dirs(base_dir, safe):
        try:
            m = os.path.getmtime(d)
        except OSError:
            continue
        if m > best_mtime:
            best_mtime = m
            best = d
    return best


def try_get_open_game_backups_folder_path(
    backup_root: str, game_name: str, subfolder_per_game: bool
) -> tuple[bool, Optional[str], Optional[str]]:
    """Return (ok, path, hint_message). Creates game subfolder when subfolder_per_game."""
    if not backup_root:
        return False, None, "No backup destination configured."
    safe = sanitize_for_windows_path_segment(game_name)
    if subfolder_per_game:
        path = os.path.join(backup_root, safe)
        try:
            os.makedirs(path, exist_ok=True)
        except OSError as exc:
            return False, None, str(exc)
        hint = None if os.path.isdir(path) else "Created backup folder for this game."
        return True, path, hint
    if not os.path.isdir(backup_root):
        try:
            os.makedirs(backup_root, exist_ok=True)
        except OSError as exc:
            return False, None, str(exc)
    return True, backup_root, None


def get_compress_candidates(
    game_names: List[str],
    backup_root: str,
    subfolder_per_game: bool,
    selected_names: Optional[List[str]] = None,
) -> List[str]:
    pool = selected_names if selected_names else game_names
    return [
        n
        for n in pool
        if n and has_retention_artifact(backup_root, n, subfolder_per_game)
    ]


def sanitized_game_names_for_filter(game_names: Iterable[str]) -> List[str]:
    return [sanitize_for_windows_path_segment(n) for n in game_names if n]
